File: tools/unpack_convert/unpack_convert.js
/* --- START OF FILE unpack_convert.js --- */
/**
 * ======================================================================
 * MetaForge Logic Bridge: Unpack & Convert
 * Build 8.2.1: Workflow Handoff Ordering Fix.
 * Role: Transitions path context to Step 2 (MusicBrainz IDs).
 * Accessibility: WCAG 2.2 AA | COGA 4.5.10 (Feedback Loops)
 * ======================================================================
 */

console.log("METAFORGE: unpack_convert.js Build 8.2.1 active...");

window.metaforge = window.metaforge || {};
window.metaforge.unpack_convert = {
    lastTrigger: null,
    selectedArtFile: null,
    observer: null,

    toTitleCase: function(str) {
        if (!str) return "";
        return str.toLowerCase().split(' ').map(word => {
            return word.charAt(0).toUpperCase() + word.slice(1);
        }).join(' ');
    },

    /**
     * Directive XIX: SPA Re-entry Protocol
     * Physically synchronizes UI with .env state using 10ms Paint Guard.
     */
    init: function() {
        const toolHeader = document.getElementById('unpack-header');
        if (!toolHeader) return;

        if (toolHeader.getAttribute('data-synced') === 'true') return;
        toolHeader.setAttribute('data-synced', 'true');

        setTimeout(async () => {
            const progLabel = document.getElementById('upk-progress-label');
            const enhancedPanel = document.getElementById('enhanced-section');
            const select = document.getElementById('upk-category');
            const consentGate = document.getElementById('upk-consent-gate');
            const fieldset = document.getElementById('upk-fieldset');

            if (!progLabel) return;
            if (fieldset) fieldset.disabled = false;
            
            try {
                const res = await fetch(`/run_tool_logic/unpack_convert/get_interview_data?t=${Date.now()}`);
                if (!res.ok) throw new Error(`HTTP ${res.status}`);
                
                const data = await res.json();
                
                if (progLabel) {
                    const formattedPolicy = this.toTitleCase(data.policy);
                    progLabel.innerText = `📚 Library Type: ${formattedPolicy}`;
                }

                if (data.policy === "ENHANCED") {
                    if (enhancedPanel) enhancedPanel.style.display = 'flex';
                    this.populateSelect(select, data.categories);
                    if (select) {
                        select.required = true;
                        select.setAttribute('aria-required', 'true');
                    }
                } else {
                    if (enhancedPanel) enhancedPanel.style.display = 'none';
                }

                if (data.stored_consent === true) {
                    if (consentGate) consentGate.style.display = 'none';
                    const consentCheckbox = document.getElementById('upk-consent');
                    if (consentCheckbox) consentCheckbox.checked = true;
                } else {
                    if (consentGate) consentGate.style.display = 'flex';
                    const consentCheckbox = document.getElementById('upk-consent');
                    if (consentCheckbox) consentCheckbox.checked = false;
                }

                this.validate();

            } catch (e) {
                console.error("METAFORGE: Sync Failure:", e);
                if (toolHeader) toolHeader.removeAttribute('data-synced');
            }
        }, 10);
    },

    setupReentryWatcher: function() {
        if (this.observer) this.observer.disconnect();

        const stage = document.getElementById('mfi-content');
        if (!stage) {
            setTimeout(() => this.setupReentryWatcher(), 100);
            return;
        }

        this.observer = new MutationObserver((mutations) => {
            for (let mutation of mutations) {
                if (mutation.addedNodes.length > 0) {
                    if (document.getElementById('unpack-header')) {
                        this.init();
                        break;
                    }
                }
            }
        });

        this.observer.observe(stage, { childList: true, subtree: false });
    },

    validate: function() {
        const btn = document.getElementById('upk-start-btn');
        const artist = document.getElementById('upk-artist')?.value.trim();
        const album = document.getElementById('upk-album')?.value.trim();
        const path = document.getElementById('upk-path')?.value.trim();
        const consent = document.getElementById('upk-consent')?.checked;
        const catIn = document.getElementById('upk-category');
        
        const isCatReady = catIn?.required ? (catIn.value !== "") : true;

        if (btn) {
            const isReady = (artist && album && path && consent && isCatReady);
            btn.disabled = !isReady;
            btn.style.opacity = isReady ? "1" : "0.5";
        }
    },

    loadArtGallery: async function() {
        const path = document.getElementById('upk-path').value;
        const consoleBox = document.getElementById('unpacker-console');
        if (!consoleBox) return;

        try {
            const res = await fetch(`/run_tool_logic/unpack_convert/get_art_gallery?path=${encodeURIComponent(path)}`);
            const images = await res.json();
            if (images.length === 0) return;

            let galleryHTML = `
                <div id="upk-step4-wrapper" style="margin-top:20px; border-top:2px solid var(--bg-accent); padding-top:15px;">
                    <div class="status-api">🖼️ Step 4: Album Cover Selection</div>
                    <div class="data-text" style="margin-bottom:10px;">Select the primary image to promote as folder.jpg</div>
                    <div id="upk-art-gallery" class="art-selection-grid">
            `;

            images.forEach((img, idx) => {
                galleryHTML += `
                    <div class="art-card" onclick="metaforge.unpack_convert.selectArt('${img.filename}', ${idx})" id="art-card-${idx}" tabindex="0" onkeydown="if(event.key==='Enter' || event.key===' ') { event.preventDefault(); metaforge.unpack_convert.selectArt('${img.filename}', ${idx}); }">
                        <div class="art-preview" style="background-image: url('${img.data}')"></div>
                        <div class="art-meta" title="${img.filename}">${img.filename}</div>
                        <div class="art-meta" style="color:var(--mf-gold)">${img.width} x ${img.height}</div>
                        <div class="art-selector-row">
                            <input type="radio" name="art-choice" class="art-radio" id="radio-${idx}" 
                                   aria-label="Select ${img.filename}, ${img.width} by ${img.height} pixels"
                                   onchange="metaforge.unpack_convert.handleRadioChange(${idx})">
                            <label for="radio-${idx}" class="art-meta">Select</label>
                        </div>
                    </div>
                `;
            });

            galleryHTML += `
                    </div>
                    <div style="text-align: right; margin-top: 10px; padding-bottom: 20px;">
                        <button id="upk-confirm-art-btn" class="mf-button-gold-fixed upk-confirm-art-btn" onclick="metaforge.unpack_convert.confirmCover()" disabled>
                            Confirm Cover
                        </button>
                    </div>
                </div>
            `;

            consoleBox.insertAdjacentHTML('beforeend', galleryHTML);
            consoleBox.scrollTop = consoleBox.scrollHeight;
        } catch (err) {
            console.error("METAFORGE: Art Gallery Error:", err);
        }
    },

    selectArt: function(filename, index) {
        this.selectedArtFile = filename;
        const btn = document.getElementById('upk-confirm-art-btn');
        if (btn) btn.disabled = false;

        document.querySelectorAll('.art-card').forEach(card => card.classList.remove('selected'));
        document.querySelectorAll('.art-radio').forEach(radio => radio.checked = false);

        const selectedCard = document.getElementById(`art-card-${index}`);
        const selectedRadio = document.getElementById(`radio-${index}`);
        if (selectedCard) selectedCard.classList.add('selected');
        if (selectedRadio) selectedRadio.checked = true;
    },

    handleRadioChange: function(index) {
        const selectedCard = document.getElementById(`art-card-${index}`);
        if (!selectedCard) return;
        const title = selectedCard.querySelector('.art-meta[title]')?.title;
        if (title) this.selectArt(title, index);
    },

    confirmCover: async function() {
        const path = document.getElementById('upk-path').value;
        const btn = document.getElementById('upk-confirm-art-btn');
        const consoleBox = document.getElementById('unpacker-console');
        const albumName = document.getElementById('upk-album').value;

        if (!this.selectedArtFile) return;
        if (btn) btn.disabled = true;

        try {
            const res = await fetch('/run_tool_logic/unpack_convert/finalize_art', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ path: path, filename: this.selectedArtFile })
            });
            const result = await res.json();

            if (result.status === "success") {
                const step4Wrapper = document.getElementById('upk-step4-wrapper');
                if (step4Wrapper) {
                    step4Wrapper.style.opacity = "0.5";
                    step4Wrapper.style.pointerEvents = "none";
                }

                const finalMessage = `
                    <div class="status-success" style="margin-top:10px; font-size:.8rem; overflow:hidden;">
                        <div>
                            Processing Complete: <span style="color:var(--mf-gold); font-weight:bold;">${albumName}</span> has been processed and is ready for the next step.
                        </div>
                    </div>
                `;
                consoleBox.insertAdjacentHTML('beforeend', finalMessage);
                this.injectHandoffButton(path);
                consoleBox.scrollTop = consoleBox.scrollHeight;
            }
        } catch (err) {
            if (consoleBox) consoleBox.insertAdjacentHTML('beforeend', `<div class="status-error">❌ Promotion Error: ${err.message}</div>`);
        } finally {
            const fieldset = document.getElementById('upk-fieldset');
            if (fieldset) fieldset.disabled = false;
        }
    },

    /**
     * Workflow Orchestration: Injects the Hand-off trigger.
     * Build 8.2.1: Logic refined to ensure button always appends to end of current log stack.
     */
    injectHandoffButton: function(passedPath) {
        const consoleBox = document.getElementById('unpacker-console');
        if (!consoleBox || document.getElementById('upk-handoff-gate')) return;

        const path = passedPath || document.getElementById('upk-path').value.trim();
        const escapedPath = path.replace(/\\/g, "\\\\");

        const handoffHTML = `
            <div id="upk-handoff-gate" style="margin-top: 20px; text-align: right; border-top: 1px solid var(--bg-accent); padding-top: 15px;">
                <button class="mf-button-gold-fixed" onclick="window.mfAdvanceWorkflow('musicbrainz_id', '${escapedPath}')">
                    Continue to: MusicBrainzIDs
                </button>
            </div>
        `;
        consoleBox.insertAdjacentHTML('beforeend', handoffHTML);
        consoleBox.scrollTop = consoleBox.scrollHeight;
    },

    updateProgress: function(phase, current, total) {
        const fill = document.getElementById('upk-progress-fill');
        const label = document.getElementById('upk-progress-label');
        const container = document.getElementById('upk-progress-container');
        
        const phaseNum = parseInt(phase) || 1;
        const curNum = parseInt(current) || 0;
        const totNum = Math.max(parseInt(total) || 1, 1);

        const phasePercent = Math.round((curNum / totNum) * 100);
        let globalPercent = (phaseNum === 1) ? (curNum / totNum) * 50 : 50 + ((curNum / totNum) * 50);
        globalPercent = Math.min(Math.max(globalPercent, 0), 100);

        const actionText = (phaseNum === 1) ? "Unpacking" : "Converting";

        if (fill) fill.style.width = `${globalPercent}%`;
        if (label) label.innerText = `${actionText} | ${curNum}/${totNum} (${phasePercent}%)`;
        if (container) container.setAttribute('aria-valuenow', Math.round(globalPercent));
    },

    populateSelect: function(select, items) {
        if (!select) return;
        const cur = select.value;
        select.innerHTML = '<option value="">-- Select Category --</option>';
        items.forEach(c => { 
            const o = document.createElement('option'); o.value = c; o.innerText = c; select.appendChild(o); 
        });
        if (cur) select.value = cur;
    },

    pickFolder: async function() {
        const path = await window.pywebview?.api?.select_folder();
        if (path) { 
            const pathIn = document.getElementById('upk-path');
            if (pathIn) pathIn.value = path; 
            this.validate(); 
        }
    },

    /**
     * Build 8.2.1 Fix: Reordered DOM update sequence to ensure triggers follow visual text.
     */
    run: async function() {
        const consoleBox = document.getElementById('unpacker-console');
        const btn = document.getElementById('upk-start-btn');
        const fieldset = document.getElementById('upk-fieldset');
        const fill = document.getElementById('upk-progress-fill');
        const label = document.getElementById('upk-progress-label');

        if (btn) btn.disabled = true;
        if (fieldset) fieldset.disabled = true;
        
        if (fill) fill.style.width = '0%';
        if (label) label.innerText = 'Starting Unpack & Convert...';
        if (consoleBox) consoleBox.innerHTML = '<div class="status-api">Initializing...</div>';
        
        const step4Wrapper = document.getElementById('upk-step4-wrapper');
        if (step4Wrapper) step4Wrapper.remove();

        const path = document.getElementById('it-path')?.value.trim() || document.getElementById('upk-path').value.trim();
        const payload = {
            artist: document.getElementById('upk-artist').value.trim(),
            album: document.getElementById('upk-album').value.trim(),
            path: path,
            category: document.getElementById('upk-category')?.value || "",
            remember: document.getElementById('upk-remember').checked,
            normalization: document.getElementById('upk-norm').checked
        };

        try {
            const res = await fetch('/run_tool_logic/unpack_convert/run_unpack', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(payload)
            });

            const reader = res.body.getReader();
            const decoder = new TextDecoder();
            let hasArtReady = false;

            while (true) {
                const { value, done } = await reader.read();
                if (done) break;
                const chunk = decoder.decode(value);
                
                // 1. Process Metadata / Non-DOM updates
                const progMatch = chunk.match(/<!-- PROGRESS:(\d+):(\d+):(\d+) -->/);
                if (progMatch) this.updateProgress(progMatch[1], progMatch[2], progMatch[3]);

                // 2. Commit chunk to DOM (This must happen first to ensure correct vertical ordering)
                if (consoleBox) {
                    consoleBox.insertAdjacentHTML('beforeend', chunk);
                }

                // 3. Evaluate Workflow Triggers based on cumulative state
                if (chunk.includes('<!-- ART_READY -->')) {
                    hasArtReady = true;
                    this.loadArtGallery();
                }

                if (chunk.includes('ready for tagging.')) {
                     this.injectHandoffButton(path);
                }
                
                if (consoleBox) {
                    consoleBox.scrollTop = consoleBox.scrollHeight;
                }
            }

            if (!hasArtReady && fieldset) {
                fieldset.disabled = false;
            }

        } catch (err) { 
            if (consoleBox) consoleBox.insertAdjacentHTML('beforeend', `<div class="status-error">FAIL: ${err.message}</div>`); 
            if (fieldset) fieldset.disabled = false;
        } finally { 
            this.validate(); 
        }
    },

    openHelp: async function() {
        const panel = document.getElementById('upk-help-panel');
        const body = document.getElementById('upk-help-body');
        const title = document.getElementById('upk-panel-title');
        this.lastTrigger = document.activeElement;
        if (!panel || !body) return;
        try {
            const response = await fetch('/tool_asset/unpack_convert/help.mfi');
            const html = await response.text();
            body.innerHTML = html;
            panel.style.display = 'flex';
            setTimeout(() => { if (title) { title.setAttribute('tabindex', '-1'); title.focus(); } }, 20);
        } catch (err) { body.innerHTML = `<p style="color:red; padding:15px;">Error: ${err.message}</p>`; panel.style.display = 'flex'; }
    },

    closeHelp: function() {
        const panel = document.getElementById('upk-help-panel');
        if (panel) { panel.style.display = 'none'; if (this.lastTrigger) this.lastTrigger.focus(); }
    }
};

window.metaforge.unpack_convert.init();
window.metaforge.unpack_convert.setupReentryWatcher();

// --- END OF FILE unpack_convert.js ---

File: tools/unpack_convert/unpack_convert.py
# --- START OF FILE unpack_convert.py ---
# ======================================================================
# MetaForge Tool Hub: Unpack & Convert
# Role: Primary tool orchestrator for ingestion and standardization.
# Build 7.9.1: Workflow Handoff Sync (Ready for Tagging Hook).
# Physical Location: \tools\unpack_convert\unpack_convert.py
# ======================================================================
import os
import sys
import re
from pathlib import Path
from flask import Response, stream_with_context, request, jsonify

# --- [ ARCHITECTURAL BOOTSTRAP ] ---
TOOL_ROOT = Path(__file__).parent.resolve()
ENGINES_DIR = TOOL_ROOT / "engines"
if str(ENGINES_DIR) not in sys.path:
    sys.path.insert(0, str(ENGINES_DIR))

import context_engine
from common import io_bridge

def run_logic(action, tools_dir, env_path):
    """
    Primary Spoke Dispatcher: Routes requests to specialized engines.
    """
    if action == "get_interview_data":
        return context_engine.handle_interview_fetch(env_path)

    if action == "run_unpack":
        data = request.json
        return Response(
            stream_with_context(_orchestrate_workflow(data, env_path)),
            mimetype='text/html'
        )

    # --- [ DISPATCHER: STEP 4 ART ROUTES ] ---
    if action == "get_art_gallery":
        path = request.args.get('path')
        if not path:
            return jsonify([])
        import art_engine
        gallery = art_engine.scan_art_folder(path)
        return jsonify(gallery)

    if action == "finalize_art":
        data = request.json
        if not data or not data.get('path') or not data.get('filename'):
            return jsonify({"status": "error", "message": "Missing required promotion data."})
        import art_engine
        result = art_engine.finalize_cover(data.get('path'), data.get('filename'))
        return jsonify(result)

    return jsonify({"status": "error", "message": f"Action {action} unrecognized."}), 400

def _natural_sort_key(text):
    """Helper for numerical sorting of directory names."""
    return [int(c) if c.isdigit() else c.lower() for c in re.split(r'(\d+)', str(text))]

def _orchestrate_workflow(data, env_path):
    """
    Orchestration Engine: Manages the bimodal lifecycle.
    Build 7.9.1: Synchronized 'ready for tagging.' hook for JS hand-off.
    """
    import cue_engine
    import zip_engine
    import rar_engine
    import sacd_engine
    import bitstream_engine
    import janitor_engine
    import art_engine

    # 1. INITIALIZATION
    root = Path(data.get('path'))
    artist = data.get('artist')
    album_name = data.get('album')
    do_norm = data.get('normalization', False)
    library_type = context_engine.get_policy(env_path)
    
    category = data.get('category')
    if library_type != "ENHANCED":
        category = None 

    report_data = {
        "phase": "Unpack & Convert",
        "extraction": {"source": "None", "count": 0},
        "conversion": [],
        "deletions": [],
        "artwork": [],
        "final_track_count": 0
    }

    if data.get('remember'):
        context_engine.update_consent_status(env_path, "TRUE")

    yield f'<div class="status-api" style="font-size:1rem;"><span aria-hidden="true">🔓</span> Starting Processing: <span style="color:var(--text-output);">{album_name}</span></div>'

    for msg in zip_engine.extract_zip(root, report_data): yield msg
    for msg in rar_engine.extract_rar(root, report_data): yield msg

    disc_pattern = re.compile(r'(?:^|[\s\-_])(disc|cd|v\.|vol|volume|part)[\s\-_]*(\d+)', re.I)
    sub_dirs = sorted([d for d in root.iterdir() if d.is_dir() and d.name.lower() != 'art'], key=lambda x: _natural_sort_key(x.name))
    disc_dirs = [d for d in sub_dirs if disc_pattern.search(d.name)]

    if disc_dirs:
        for d_path in disc_dirs:
            for msg in zip_engine.extract_zip(d_path, report_data): yield msg
            for msg in rar_engine.extract_rar(d_path, report_data): yield msg
            for msg in sacd_engine.process_sacd(d_path, artist, report_data): yield msg
            for msg in cue_engine.split_cue(d_path, artist, report_data): yield msg
    else:
        for msg in sacd_engine.process_sacd(root, artist, report_data): yield msg
        for msg in cue_engine.split_cue(root, artist, report_data): yield msg

    if not report_data.get('extraction_occurred', False):
        yield '<div class="status-api" style="margin-top:5px; padding-top:5px;"><span aria-hidden="true">📦</span> Step 1: Inspecting Original Audio Source. <br><span style="margin-left:25px; color:var(--text-output)">No archive bundles found. Unpacking not required.</span></div>'
        yield "<!-- PROGRESS:1:1:1 -->"

    # 3. STEP 2: BITSTREAM & METADATA
    if disc_dirs:
        total_discs = len(disc_dirs)
        linear_offset = 0
        yield f'<div class="status-message" style="margin-left:1.5rem; margin-top:5px; display:flex; align-items:center; gap:8px;"><img src="/ui/images/multi-disc.svg" alt="" aria-hidden="true" style="margin-left:3px; height:20px; width:auto;"> Box Set Detected: <strong>{total_discs} volume(s)</strong> found.</div>'
        
        for d_idx, d_path in enumerate(disc_dirs, 1):
            for msg in bitstream_engine.process_targets(d_path, artist, album_name, category, report_data, 
                                                       disc_num=d_idx, 
                                                       total_discs=total_discs, 
                                                       offset=linear_offset, 
                                                       is_multidisc=True,
                                                       do_norm=do_norm):
                yield msg
            
            audio_exts = {'.mp3', '.flac', '.wav', '.ape', '.m4a', '.ogg'}
            linear_offset += len([f for f in d_path.iterdir() if f.suffix.lower() in audio_exts])
    else:
        for msg in bitstream_engine.process_targets(root, artist, album_name, category, report_data, 
                                                   disc_num=1, 
                                                   total_discs=1, 
                                                   offset=0, 
                                                   is_multidisc=False,
                                                   do_norm=do_norm):
            yield msg

    # 4. STEP 3: JANITORIAL PHASE
    yield from janitor_engine.execute_cleanup(root, report_data)

    # 5. STEP 4: SIGNAL ART SELECTION OR ACTIVATE CLOUD FALLBACK
    if report_data.get('artwork') and len(report_data['artwork']) > 0:
        # JS picks this up and shows the gallery
        yield "<!-- ART_READY -->"
    else:
        yield f'<div class="status-api" style="margin-top:5px; padding-top:10px;"><span aria-hidden="true">🖼️</span> Step 4: Album Cover Selection: Local artwork not found.</div>'
        yield f'<div class="status-message" style="color:var(--text-output); margin-left:1rem; margin-left:15px;"><img src="/ui/images/cloud.png" style="height:14px; width:auto;" alt=""> Querying Discogs Archive for {album_name}...</div>'
        
        cloud_res = art_engine.trigger_discogs_fetch(root, artist, album_name)
        
        if cloud_res['status'] == 'success':
            # Build 7.9.1: Inclusion of "ready for tagging." ensures injectHandoffButton triggers in JS
            yield f'''
            <div class="status-success" style="margin-top:10px; font-size:.8rem; overflow:hidden;">
                <div>
                    Cloud Recovery Successful: Archival cover fetched from <strong>{cloud_res['source']}</strong>.<br>
                    <span style="color:var(--text-message); font-size:0.7rem;">{cloud_res['embedded_count']} files updated with archival artwork ({cloud_res['dims']}).</span><br>
                    <span style="color:var(--mf-gold); font-weight:bold;">{album_name}</span> is now prepped and ready for tagging.
                </div>
            </div>'''
        else:
            # Build 7.9.1: Inclusion of "ready for tagging." ensures injectHandoffButton triggers in JS
            yield f'''
            <div class="status-success" style="margin-top:10px; font-size:.8rem; overflow:hidden;">
                <div>
                    Processing Complete: <span style="color:var(--mf-gold); font-weight:bold;">{album_name}</span> is now prepped and ready for tagging.<br>
                    <span style="color:var(--status-error); font-size:0.7rem;">(Note: No artwork identified locally or in cloud archives)</span>
                </div>
            </div>'''

    # 6. FINALIZATION & FORENSIC AUDIT
    final_audio_list = io_bridge.get_audio_targets(root, recursive=True)
    report_data["final_track_count"] = len(final_audio_list)
    
    context_engine.write_audit_log(root, report_data)
    
    manifest_category = category if category else "None Provided"
    context_engine.generate_manifest(
        root, 
        artist, 
        album_name, 
        manifest_category, 
        library_type, 
        track_count=report_data["final_track_count"]
    )

# --- END OF FILE unpack_convert.py ---